File: algorithms/code/leetcode/lc038_count_and_say/LC038CountAndSay.java
package leetcode.lc038_count_and_say;

/**
 * https://leetcode.com/problems/count-and-say/
 * #medium
 */
public final class LC038CountAndSay {
    public String countAndSay(final int nValue) {
        int n = nValue;
        if (n < 1) {
            return null;
        }
        String result = "1";
        while (n > 1) {
            StringBuilder current = new StringBuilder();
            char check = '0';
            int count = 0;
            for (char character : result.toCharArray()) {
                if (check != character) {
                    if (count > 0) {
                        current.append(count);
                        current.append(check);
                    }
                    count = 1;
                    check = character;
                } else {
                    count++;
                }
            }
            current.append(count);
            current.append(check);
            n--;
            result = current.toString();
        }
        return result;
    }
}

File: algorithms/code/leetcode/lc038_count_and_say/lc038_count_and_say.rs
// https://leetcode.com/problems/count-and-say/
// #medium

pub fn count_and_say(n_value: i32) -> String {
    let mut n = n_value;
    if n < 1 {
        return "".to_string();
    }
    let mut result = "1".to_string();
    while n > 1 {
        let mut current = String::new();
        let mut check: char = '\x00';
        let mut count = 0;
        for character in result.chars() {
            if character != check {
                if count > 0 {
                    current.push_str(&count.to_string());
                    current.push(check);
                }
                count = 1;
                check = character;
            } else {
                count += 1;
            }
        }
        current.push_str(&count.to_string());
        current.push(check);
        result = current;
        n -= 1;
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test1() {
        assert_eq!(count_and_say(1), "1".to_string());
    }

    #[test]
    fn test2() {
        assert_eq!(count_and_say(2), "11".to_string());
    }

    #[test]
    fn test3() {
        assert_eq!(count_and_say(3), "21".to_string());
    }

    #[test]
    fn test4() {
        assert_eq!(count_and_say(4), "1211".to_string());
    }

    #[test]
    fn test5() {
        assert_eq!(count_and_say(5), "111221".to_string());
    }

    #[test]
    fn test0() {
        assert!(count_and_say(0).is_empty());
    }
}
